File: app/src/main/java/com/zack/enderplan/view/fragment/SettingsFragment.java
package com.zack.enderplan.view.fragment;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.app.AppCompatDelegate;
import android.support.v7.preference.ListPreference;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceFragmentCompat;
import android.support.v7.preference.SwitchPreferenceCompat;

import com.zack.enderplan.R;
import com.zack.enderplan.common.Constant;
import com.zack.enderplan.model.DataManager;
import com.zack.enderplan.model.preference.PreferenceHelper;
import com.zack.enderplan.view.activity.HomeActivity;
import com.zack.enderplan.view.dialog.MessageDialogFragment;

public class SettingsFragment extends PreferenceFragmentCompat implements SharedPreferences.OnSharedPreferenceChangeListener {

    private DataManager mDataManager = DataManager.getInstance();

    private SwitchPreferenceCompat mNightModePreference;
    private ListPreference mDrawerHeaderDisplayPreference;
    private ListPreference mTypeListItemEndDisplayPreference;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.preferences);

        mNightModePreference = (SwitchPreferenceCompat) findPreference(Constant.PREF_KEY_NIGHT_MODE);
        mDrawerHeaderDisplayPreference = (ListPreference) findPreference(Constant.PREF_KEY_DRAWER_HEADER_DISPLAY);
        mTypeListItemEndDisplayPreference = (ListPreference) findPreference(Constant.PREF_KEY_TYPE_LIST_ITEM_END_DISPLAY);

        mNightModePreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                MessageDialogFragment fragment = MessageDialogFragment.newInstance(getString(R.string.title_dialog_switch_night_mode), getString(R.string.msg_dialog_switch_night_mode), null, getString(R.string.button_cancel), getString(R.string.button_restart));
                fragment.setOnPositiveButtonClickListener(new MessageDialogFragment.OnPositiveButtonClickListener() {
                    @Override
                    public void onPositiveButtonClick() {
                        mNightModePreference.setChecked(!mNightModePreference.isChecked());
                    }
                });
                fragment.show(getFragmentManager());
                //返回false表示不改变preference的值
                return false;
            }
        });

        setSummary(Constant.PREF_KEY_DRAWER_HEADER_DISPLAY);
        setSummary(Constant.PREF_KEY_TYPE_LIST_ITEM_END_DISPLAY);
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        switch (key) {
            case Constant.PREF_KEY_NIGHT_MODE:
                AppCompatDelegate.setDefaultNightMode(mDataManager.getPreferenceHelper().getNightModeValue() ? AppCompatDelegate.MODE_NIGHT_YES : AppCompatDelegate.MODE_NIGHT_NO);
                //返回并重新创建HomeActivity
                HomeActivity.start(getActivity());
                break;
            case Constant.PREF_KEY_DRAWER_HEADER_DISPLAY:
                setSummary(Constant.PREF_KEY_DRAWER_HEADER_DISPLAY);
                break;
            case Constant.PREF_KEY_TYPE_LIST_ITEM_END_DISPLAY:
                setSummary(Constant.PREF_KEY_TYPE_LIST_ITEM_END_DISPLAY);
                break;
        }
    }

    private void setSummary(String key) {
        PreferenceHelper preferenceHelper = mDataManager.getPreferenceHelper();
        switch (key) {
            case Constant.PREF_KEY_DRAWER_HEADER_DISPLAY:
                mDrawerHeaderDisplayPreference.setSummary(mDrawerHeaderDisplayPreference.getEntries()[mDrawerHeaderDisplayPreference.findIndexOfValue(preferenceHelper.getDrawerHeaderDisplayValue())]);
                break;
            case Constant.PREF_KEY_TYPE_LIST_ITEM_END_DISPLAY:
                mTypeListItemEndDisplayPreference.setSummary(mTypeListItemEndDisplayPreference.getEntries()[mTypeListItemEndDisplayPreference.findIndexOfValue(preferenceHelper.getTypeListItemEndDisplayValue())]);
                break;
        }
    }
}

File: app/src/main/java/net/zackzhang/app/end/view/fragment/SettingsFragment.kt
package net.zackzhang.app.end.view.fragment

import android.content.SharedPreferences
import android.os.Bundle
import android.support.v14.preference.SwitchPreference
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatDelegate
import android.support.v7.preference.ListPreference
import android.support.v7.preference.Preference
import android.support.v7.preference.PreferenceFragmentCompat
import net.zackzhang.app.end.R
import net.zackzhang.app.end.common.Constant
import net.zackzhang.app.end.model.DataManager
import net.zackzhang.app.end.view.activity.HomeActivity
import net.zackzhang.app.end.view.dialog.MessageDialogFragment

class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {

    companion object {

        private const val TAG_SWITCH_NIGHT_MODE = "switch_night_mode"
    }

    private val nightModePreference by lazy {
        findPreference(Constant.PREF_KEY_NIGHT_MODE) as SwitchPreference
    }
    // TODO ListPreferenceDialogFragment
    private val drawerHeaderDisplayPreference by lazy {
        findPreference(Constant.PREF_KEY_DRAWER_HEADER_DISPLAY) as ListPreference
    }
    private val typeListItemEndDisplayPreference by lazy {
        findPreference(Constant.PREF_KEY_TYPE_LIST_ITEM_END_DISPLAY) as ListPreference
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.preferences)

        nightModePreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, _ ->
            MessageDialogFragment.Builder()
                    .setTitle(R.string.title_dialog_switch_night_mode)
                    .setMessage(R.string.msg_dialog_switch_night_mode)
                    .setOkButtonText(R.string.button_restart)
                    .showCancelButton()
                    .show(childFragmentManager, TAG_SWITCH_NIGHT_MODE)
            // 返回 false 表示不改变 preference 的值
            false
        }

        setSummary(Constant.PREF_KEY_DRAWER_HEADER_DISPLAY)
        setSummary(Constant.PREF_KEY_TYPE_LIST_ITEM_END_DISPLAY)
    }

    override fun onResume() {
        super.onResume()
        preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onAttachFragment(childFragment: Fragment) {
        super.onAttachFragment(childFragment)

        if (childFragment.tag == TAG_SWITCH_NIGHT_MODE) {
            (childFragment as MessageDialogFragment).okButtonClickListener = { nightModePreference.isChecked = !nightModePreference.isChecked }
        }
    }

    override fun onPause() {
        super.onPause()
        preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        when (key) {
            Constant.PREF_KEY_NIGHT_MODE -> {
                AppCompatDelegate.setDefaultNightMode(if (DataManager.preferenceHelper.nightModeValue) AppCompatDelegate.MODE_NIGHT_YES else AppCompatDelegate.MODE_NIGHT_NO)
                // 返回并重新创建 HomeActivity
                HomeActivity.start(context!!)
            }
            Constant.PREF_KEY_DRAWER_HEADER_DISPLAY -> setSummary(Constant.PREF_KEY_DRAWER_HEADER_DISPLAY)
            Constant.PREF_KEY_TYPE_LIST_ITEM_END_DISPLAY -> setSummary(Constant.PREF_KEY_TYPE_LIST_ITEM_END_DISPLAY)
        }
    }

    private fun setSummary(key: String) {
        val preferenceHelper = DataManager.preferenceHelper
        when (key) {
            Constant.PREF_KEY_DRAWER_HEADER_DISPLAY -> drawerHeaderDisplayPreference.summary = drawerHeaderDisplayPreference.entries[drawerHeaderDisplayPreference.findIndexOfValue(preferenceHelper.drawerHeaderDisplayValue)]
            Constant.PREF_KEY_TYPE_LIST_ITEM_END_DISPLAY -> typeListItemEndDisplayPreference.summary = typeListItemEndDisplayPreference.entries[typeListItemEndDisplayPreference.findIndexOfValue(preferenceHelper.typeListItemEndDisplayValue)]
        }
    }
}
